modules/renter/hostdb/negotiate.go

```diff
@@ -220,304 +119,207 @@ func (hu *hostUploader) negotiateContract(filesize uint64, duration types.BlockH
 	signedTxnSet, err := txnBuilder.Sign(true)
 	if err != nil {
 		txnBuilder.Drop()
-		return err
+		return hostContract{}, err
 	}
 	if err := encoding.WriteObject(conn, signedTxnSet); err != nil {
 		txnBuilder.Drop()
-		return errors.New("couldn't send the contract signed by us: " + err.Error())
+		return hostContract{}, errors.New("couldn't send the contract signed by us: " + err.Error())
 	}
 
 	// read signed txn from host
 	var signedHostTxnSet []types.Transaction
 	if err := encoding.ReadObject(conn, &signedHostTxnSet, types.BlockSizeLimit); err != nil {
 		txnBuilder.Drop()
-		return errors.New("couldn't read the contract signed by the host: " + err.Error())
+		return hostContract{}, errors.New("couldn't read the contract signed by the host: " + err.Error())
 	}
 
 	// submit to blockchain
-	err = hu.hdb.tpool.AcceptTransactionSet(signedHostTxnSet)
+	err = tpool.AcceptTransactionSet(signedHostTxnSet)
 	if err == modules.ErrDuplicateTransactionSet {
 		// this can happen if the renter is uploading to itself
 		err = nil
 	}
 	if err != nil {
 		txnBuilder.Drop()
-		return err
-	}
-
-	hu.fcid = fcid
-	// create initial revision
-	hu.lastTxn.FileContractRevisions = []types.FileContractRevision{{
-		ParentID:              fcid,
-		UnlockConditions:      hu.unlockConditions,
-		NewRevisionNumber:     fc.RevisionNumber,
-		NewFileSize:           fc.FileSize,
-		NewFileMerkleRoot:     fc.FileMerkleRoot,
-		NewWindowStart:        fc.WindowStart,
-		NewWindowEnd:          fc.WindowEnd,
-		NewValidProofOutputs:  []types.SiacoinOutput{fc.ValidProofOutputs[0], fc.ValidProofOutputs[1]},
-		NewMissedProofOutputs: []types.SiacoinOutput{fc.MissedProofOutputs[0], fc.MissedProofOutputs[1]},
-		NewUnlockHash:         fc.UnlockHash,
-	}}
-
-	hu.hdb.mu.Lock()
-	hu.hdb.contracts[fcid] = hostContract{
+		return hostContract{}, err
+	}
+
+	// create host contract
+	hc := hostContract{
+		IP:           addr,
 		ID:           fcid,
 		FileContract: fc,
 		LastRevisionTxn: types.Transaction{
-			// first revision is empty
-			FileContractRevisions: []types.FileContractRevision{{}},
+			FileContractRevisions: []types.FileContractRevision{{
+				ParentID:              fcid,
+				UnlockConditions:      uc,
+				NewRevisionNumber:     fc.RevisionNumber,
+				NewFileSize:           fc.FileSize,
+				NewFileMerkleRoot:     fc.FileMerkleRoot,
+				NewWindowStart:        fc.WindowStart,
+				NewWindowEnd:          fc.WindowEnd,
+				NewValidProofOutputs:  []types.SiacoinOutput{fc.ValidProofOutputs[0], fc.ValidProofOutputs[1]},
+				NewMissedProofOutputs: []types.SiacoinOutput{fc.MissedProofOutputs[0], fc.MissedProofOutputs[1]},
+				NewUnlockHash:         fc.UnlockHash,
+			}},
 		},
```

Comment:
So, this is one of the ugly constructions that remains. When you revise, you want to revise the previous revision. But after the initial contract negotiation, there are no revisions yet. So a dummy "0th revision" is created for future code to build off of.
Another related bit of ugliness is that you have to reference the revision with `contract.LastRevisionTxn.FileContractRevisions[0]`, which is both verbose and risks an out-of-bounds error. One obvious way to rectify this would be to add a `LastRevision` field to the `hostContract` type, but then you're duplicating data and you have to be careful that it stays synchronized with the revision in `LastRevisionTxn`.